lib/evaluate/fold.cc

```diff
@@ -152,35 +157,486 @@ ComplexPart FoldOperation(FoldingContext &context, ComplexPart &&complexPart) {
       FoldOperation(context, std::move(complex)), complexPart.part()};
 }
 
+// Scalar conversion
+// Conversion Helper to be used to avoid instantiating ScalarConvert when
+// it would be undefined.
+template<TypeCategory TOCAT, TypeCategory FROMCAT>
+static constexpr inline bool IsAllowedBasicConversion() {
```

Comment:
Does Basic here correspond to conversions of KIND?

Reply:
No it means that it is allowed to have a `Convert<TO, FROMCAT>` with `TO::category = TOCAT`.
Technically, it is allowed to convert a Complex to Real or Integer or vice-versa, but `Convert` expression do not directly accept Complex types, instead, Complex conversions are done piece-wise using `ComplexComponent` to split before the `Convert` and `ComplexConstructor` to recompose after.
"Basic" was supposed to convey this, I agree it deserves a better name. Do you think `IsAllowedTypesInsideConvertExpr` is better?

Reply:
Ah, "basic" would've made perfect sense in a C context.  In Fortran, not as much.  Maybe something like `IsAllowedIntrinsicTypeConversion()` would be more Fortran-ish.